cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -738,21 +740,37 @@ public ApplicationSpecification apply(ApplicationMeta input) {
   }
 
   @Override
-  public void addSchedule(final ProgramId program, final ScheduleSpecification scheduleSpecification) {
-    Transactions.executeUnchecked(transactional, new TxRunnable() {
-      @Override
-      public void run(DatasetContext context) throws Exception {
-        AppMetadataStore metaStore = getAppMetadataStore(context);
-        ApplicationSpecification appSpec = getAppSpecOrFail(metaStore, program);
-        Map<String, ScheduleSpecification> schedules = Maps.newHashMap(appSpec.getSchedules());
-        String scheduleName = scheduleSpecification.getSchedule().getName();
-        Preconditions.checkArgument(!schedules.containsKey(scheduleName), "Schedule with the name '" +
-          scheduleName + "' already exists.");
-        schedules.put(scheduleSpecification.getSchedule().getName(), scheduleSpecification);
-        ApplicationSpecification newAppSpec = new AppSpecificationWithChangedSchedules(appSpec, schedules);
-        metaStore.updateAppSpec(program.getNamespace(), program.getApplication(), program.getVersion(), newAppSpec);
+  public void addSchedule(final ProgramId program, final ScheduleSpecification scheduleSpecification,
+                          final boolean allowOverwrite) throws AlreadyExistsException {
+    try {
+      Transactions.executeUnchecked(transactional, new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          AppMetadataStore metaStore = getAppMetadataStore(context);
+          ApplicationSpecification appSpec = getAppSpecOrFail(metaStore, program);
+          Map<String, ScheduleSpecification> existingSchedules = appSpec.getSchedules();
+          String scheduleName = scheduleSpecification.getSchedule().getName();
+          if (!allowOverwrite && existingSchedules.containsKey(scheduleName)) {
+            throw new AlreadyExistsException(
+              new ScheduleId(program.getNamespace(), program.getApplication(), scheduleName));
+          }
+
+          Map<String, ScheduleSpecification> schedules = Maps.newHashMap(existingSchedules);
+          schedules.put(scheduleSpecification.getSchedule().getName(), scheduleSpecification);
+          ApplicationSpecification newAppSpec = new AppSpecificationWithChangedSchedules(appSpec, schedules);
+          metaStore.updateAppSpec(program.getNamespace(), program.getApplication(), program.getVersion(), newAppSpec);
+          LOG.debug("{} schedule for program {} - {}",
+                    existingSchedules.containsKey(scheduleName) ? "Added" : "Updated", program, scheduleSpecification);
+        }
+      });
+    } catch (Exception e) {
+      // Transactions.executeUnchecked wraps all exceptions in RuntimeException
+      if (e.getCause() instanceof AlreadyExistsException) {
```

Comment:
have you tested this? I think that executeUnchecked will throw AlreadyExistsException, wrapped in TransactionFailureException, wrapped in RuntimeException.